proofs/pseudocode-defns/pseudocode_defns.tex

```diff
@@ -0,0 +1,797 @@
+\end{lstlisting}
+    
+\subsection{Pseudocode without preconditions}
+\label{sec:pseudocode-wout-preconditions}
+\begin{lstlisting}[language=Python]
+class Measurement:
+    def __init__(self, input_domain, output_domain, function, input_metric, output_measure, privacy_map):
+        
+        assert has_trait(input_domain, Domain)
+        self.input_domain = input_domain
+        assert has_trait(output_domain, Domain)
+        self.output_domain = output_domain
+        
+        assert get_input_domain(function) == input_domain
+        assert get_output_domain(function) == output_domain
+        self.function = function
+        
+        assert has_trait(input_metric, Metric)
+        self.input_metric = input_metric
+        assert has_trait(output_measure, Measure)
+        self.output_measure = output_measure
+        
+        assert get_input_metric(privacy_relation) == input_metric
+        assert get_output_measure(privacy_relation) == output_measure
+        self.privacy_map = privacy_map
+\end{lstlisting}
+    
+    \inOpenDPRust{Measurement}{https://github.com/opendp/opendp/blob/35dbdc73d7d74e049f5101a704d4e036bed365e8/rust/opendp/src/core.rs/#L339-L346} When we refer to a \textit{valid measurement} in the proofs, this is the precise definition.
+\end{definition}
+Therefore, there is no need to include the following code snippet in all of the pseudocodes:
+\begin{lstlisting}[language=Python]
+class Measurement:
+    input_domain
+    output_domain
+    function
+    input_metric
+    output_measure
+    privacy_map 
+\end{lstlisting}
+\section{Metrics}
+\label{sec:metrics}
+\subsection{Dataset metrics}
+Metrics are used to measure the distances between data. Metrics have a \emph{domain} on which the function associated with the metric can measure distance, and an \emph{associated type} that determines the type used to represent the distance between datasets.
+\textbf{Example:} \texttt{SymmetricDistance} has a domain of \texttt{VectorDomain(AllDomain(T))}, which means that \texttt{SymmetricDistance} can be used to measure the distance between any objects that are vectors of elements of type \texttt{T}. \texttt{SymmetricDistance} has an associated type of \texttt{u32}, which means that a \texttt{u32} value is used to report the distance.
+Please refer to the proof definitions document (see section \ref{sec:versioned-docs}) for definitions of the functions referred to by the metrics below.
+\note{Once the subdomains have been implemented in OpenDP, it will no longer be necessary to list all of the subdomains in the compatible domains section.}
+\begin{definition}[\texttt{AbsoluteDistance(T)}]
+\metricDefn{absolute distance}
+\begin{itemize}
+    \item \textbf{Domain:} \texttt{AllDomain(T)}, where \texttt{T} has the traits \texttt{Sub(Output=T)} (defined in definition \ref{defn:trait-sub}) and \texttt{TotalOrd} (defined in definition \ref{defn:trait-totalord}).
+    \item \textbf{Associated type:} \texttt{Q}.
```

Comment:
Here Q is not quantified/defined.  I understand that this Q should be a T.  Should also check in the definitions below.